docs: add using strong-error-handler page (#5392)
strongloop/loopback-next: docs/site/Using-strong-error-handler.md

---
title: 'Using strong-error-handler'
lang: en
keywords: LoopBack 4, error handling
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Using-strong-error-handler.html
---

# strong-error-handler

As the part of the dependencies of
[`@loopback/rest`](https://github.com/strongloop/loopback-next/tree/master/packages/rest),
package `strong-error-handler` is an error handler for use in both development
(debug) and production environments.

In LoopBack 4, the request handling process starts with the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html), a simple class with
five injected helper methods in the constructor. It is the gatekeeper of all
requests to the app. Errors are handled by one of the Sequence actions,
[`reject`](Sequence.md#handling-errors), which calls `strong-error-handler`
package to send back an HTTP response describing the error.

LB4 apps require 3.x versions of `strong-error-handler`.

## Usage

As mentioned above, errors are handled by the `reject` action of the app's

> [sequence](https://loopback.io/doc/en/lb4/Sequence.html). By default, `reject`

Reviewer: Nitpick: a suggestion for rephrasing the usage example :)

---
You can configure the error handler's behaviors by setting the error writer options.
For example, by default, `reject` omits details from error responses to prevent leaking sensitive information. But `reject` also can return full error stack traces and internal details of any error objects to the client in the HTTP responses. This is controlled by the option property `debug`:

```ts
app.bind(RestBindings.ERROR_WRITER_OPTIONS).to({debug: true});
```

Once the `debug` flag is set, `strong-error-handler` would handle such options.

---
title: 'Using strong-error-handler'
lang: en
keywords: LoopBack 4, error handling
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Using-strong-error-handler.html
---

# strong-error-handler

> As the part of the dependencies of

Reviewer 2: `As a dependency of`?

[`@loopback/rest`](https://github.com/strongloop/loopback-next/tree/master/packages/rest),
package `strong-error-handler` is an error handler for use in both development
(debug) and production environments.

In LoopBack 4, the request handling process starts with the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html), a simple class with
five injected helper methods in the constructor. It is the gatekeeper of all
requests to the app. Errors are handled by one of the Sequence actions,
[`reject`](Sequence.md#handling-errors), which calls `strong-error-handler`
package to send back an HTTP response describing the error.

> LB4 apps require 3.x versions of `strong-error-handler`.

Reviewer 2: `LB4` -> `LoopBack 4`.

---
title: 'Using strong-error-handler'
lang: en
keywords: LoopBack 4, error handling
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Using-strong-error-handler.html
---

# strong-error-handler

This package is an error handler for use in both development (debug) and
production environments.

In LoopBack 4, the request handling process starts with the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html), a simple class with
five injected helper methods in the constructor. It is the gatekeeper of all
requests to the app. Errors are handled by one of the Sequence actions,
[`reject`](Sequence.md#handling-errors), which calls `strong-error-handler`
package to send back an HTTP response describing the error.

LB4 apps require 3.x versions of `strong-error-handler`.

## Installation

```bash

> $ npm install --save strong-error-handler

Reviewer 2: For LoopBack's `@loopback/rest` package, `strong-error-handler` is already a dependency. I don't think an explicit installation is needed.

Reviewer 3: +1

```

## Usage

As mentioned above, errors are handled by the `reject` action of the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html). By default,
implementation of `reject`, `strong-error-handler` omits details from error
responses to prevent leaking sensitive information.

`strong-error-handler` also can return full error stack traces and internal
details of any error objects to the client in the HTTP responses once the
`debug` flag is set. Please check
[Sequence - Handling errors](Sequence.md#handling-errors) section for usages and
examples.

In general, `strong-error-handler` must be the last middleware function to be

> registered.

Reviewer 2: For the sequence we have, the `reject` is executed in `try {} catch(err) {}` block. `strong-error-handler` itself is a utility, not a middleware.

The above configuration will log errors to the server console, but not return
stack traces in HTTP responses. For details on configuration options, see below.

### Working with Express middleware

Under the hood, LoopBack leverages [Express](https://expressjs.com) framework
and its concept of middleware. To avoid common pitfalls, it is not possible to

> mount Express middleware directly on a LoopBack application.

Reviewer 2: This has changed since the introduction of Express middleware support in `@loopback/rest@4.0.0`. But the key is that the error handler is stilled used.

---
title: 'Using strong-error-handler'
lang: en
keywords: LoopBack 4, error handling
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Using-strong-error-handler.html
---

# strong-error-handler

This package is an error handler for use in both development (debug) and
production environments.

In LoopBack 4, the request handling process starts with the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html), a simple class with
five injected helper methods in the constructor. It is the gatekeeper of all
requests to the app. Errors are handled by one of the Sequence actions,
[`reject`](Sequence.md#handling-errors), which calls `strong-error-handler`

> package.

Reviewer: ```suggestion
package to send back an HTTP response describing the error.
```
It'd be better to explain calls the module for what purpose.

LB4 apps require 3.x versions of `strong-error-handler`.

## Installation

```bash
$ npm install --save strong-error-handler
```

## Usage

As mentioned above, errors are handled by the `reject` action of the app's

> [sequence](https://loopback.io/doc/en/lb4/Sequence.html). By default,

Author: I'd like to explain that in LB4, errors are handled by `Sequence.reject` first, then explain the default mode and the debug mode. Do you have any suggestions of rewording it?

implementation of `reject`, `strong-error-handler` omits details from error
responses to prevent leaking sensitive information.

`strong-error-handler` also can return full error stack traces and internal
details of any error objects to the client in the HTTP responses once the
`debug` flag is set. Please check
[Sequence - Handling errors](Sequence.md#handling-errors) section for usages and
examples.

The module also exports `writeErrorToResponse`, a non-middleware flavor of the
error handler:

```js
import {Server} from 'http';
import {writeErrorToResponse} from 'strong-error-handler';
const errHandlingOptions = {debug: true, log: true};

> http

Reviewer: could you elaborate more about how to use/combine this in a lB4 application?

Author: I decide to remove this part as I think it's the same as the [Customizing log message] section below.

  .createServer(function handleRequest(req, res) {
    if (errShouldBeThrown) {
      writeErrorToResponse(
        new Error('something went wrong'),
        req,
        res,
        errHandlingOptions,
      );
    }
  })
  .listen(3000);
```

In general, `strong-error-handler` must be the last middleware function to be
registered.

The above configuration will log errors to the server console, but not return
stack traces in HTTP responses. For details on configuration options, see below.

### Working with Express middleware

Under the hood, LoopBack leverages [Express](https://expressjs.com) framework
and its concept of middleware. To avoid common pitfalls, it is not possible to
mount Express middleware directly on a LoopBack application.

`strong-error-handler` can be bond to the app, which is the same as above. For
example:

```ts

> app.bind(RestBindings.ERROR_WRITER_OPTIONS).to({debug: true});

Reviewer: > `strong-error-handler` can be bond to the app, which is the same as above

The binding here only adds the configuration, could you add the code showing how to bind the handler itself?

### Working with Express middleware

Under the hood, LoopBack leverages [Express](https://expressjs.com) framework
and its concept of middleware. To avoid common pitfalls, it is not possible to
mount Express middleware directly on a LoopBack application.

`strong-error-handler` can be bond to the app, which is the same as above. For
example:

```ts
app.bind(RestBindings.ERROR_WRITER_OPTIONS).to({debug: true});
```

### Response format and content type

The `strong-error-handler` package supports JSON, HTML and XML responses:

- When the object is a standard Error object, it returns the string provided by
  the stack property in HTML/text responses.
- When the object is a non-Error object, it returns the result of `util.inspect`
  in HTML/text responses.
- For JSON responses, the result is an object with all enumerable properties
  from the object in the response.

The content type of the response depends on the request's `Accepts` header.

- For Accepts header `json` or `application/json`, the response content type is
  JSON.
- For Accepts header `html` or `text/html`, the response content type is HTML.
- For Accepts header `xml` or `text/xml`, the response content type is XML.

> _There are plans to support other formats such as text/plain._

Reviewer 4: nitpick
```suggestion
_There are plans to support other formats such as `text/plain`._
```

---
title: 'Using strong-error-handler'
lang: en
keywords: LoopBack 4, error handling
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Using-strong-error-handler.html
summary:
  Strong-error-handler is an error handler for use in both development and
  production environments.
---

In LoopBack 4, the request handling process starts with the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html), a simple class with
five injected helper methods in the constructor. It is the gatekeeper of all
requests to the app. Errors are handled by one of the Sequence actions,
[`reject`](Sequence.md#handling-errors), which calls `strong-error-handler`
package.

# strong-error-handler

This package is an error handler for use in both development (debug) and
production environments.

> ## Supported versions

Reviewer 4: just a suggestion.. it seems a bit strange to have this supported version table where there's only one entry. If only 3.x works with LB4, perhaps we can mention in the paragraph above that LB4 requires `strong-error-handler` 3.x and up? (though there's no "up" yet).

Author: Sounds good! I agree that it's weird to have a table that only have one item :sweat_smile: Will reword it.

## Installation

```bash
$ npm install --save strong-error-handler
```

## Usage

As mentioned above, errors are handled by the `reject` action of the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html). By default,
implementation of `reject`, `strong-error-handler` omits details from error
responses to prevent leaking sensitive information.

`strong-error-handler` also can return full error stack traces and internal
details of any error objects to the client in the HTTP responses once the
`debug` flag is set. Please check
[Sequence - Handling errors](Sequence.md#handling-errors) section for usages and
examples.

The module also exports `writeErrorToResponse`, a non-middleware flavor of the
error handler:

```js
import {Server} from 'http';
import {writeErrorToResponse} from 'strong-error-handler';
const errHandlingOptions = {debug: true, log: true};

http
  .createServer(function handleRequest(req, res) {
    if (errShouldBeThrown) {
      writeErrorToResponse(
        new Error('something went wrong'),
        req,
        res,
        errHandlingOptions,
      );
    }
  })
  .listen(3000);
```

In general, `strong-error-handler` must be the last middleware function

> registered.

Reviewer 4: maybe this is more grammatically correct? (you may want to double check) 
```suggestion
to be registered.
```

### Working with Express middleware

Under the hood, LoopBack leverages [Express](https://expressjs.com) framework
and its concept of middleware. To avoid common pitfalls, it is not possible to
mount Express middleware directly on a LoopBack application.

`strong-error-handler` can be bond to the app, which is the same as above. For
example:

```ts
app.bind(RestBindings.ERROR_WRITER_OPTIONS).to({debug: true});
```

### Response format and content type

The `strong-error-handler` package supports JSON, HTML and XML responses:

- When the object is a standard Error object, it returns the string provided by
  the stack property in HTML/text responses.
- When the object is a non-Error object, it returns the result of `util.inspect`
  in HTML/text responses.
- For JSON responses, the result is an object with all enumerable properties
  from the object in the response.

The content type of the response depends on the request's `Accepts` header.

- For Accepts header `json` or `application/json`, the response content type is
  JSON.
- For Accepts header `html` or `text/html`, the response content type is HTML.
- For Accepts header `xml` or `text/xml`, the response content type is XML.

> _There are plans to support other formats such as Plain-text._

Reviewer 4: should it be `text/plain`? 
See https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types.

          code: 'VALIDATION_FAILED',
          ...customizedProps,
        };

        context.response.status(422).send(errorData);

        // 4. log the error using RestBindings.SequenceActions.LOG_ERROR
        this.logError(err, err.statusCode, context.request);

        // The error was handled
        return;
      }
    }

    // Otherwise fall back to the default error handler
    this.reject(context, err);
  }
}
```

### Safe error fields

By default, `strong-error-handler` will only pass through the `name`, `message`
and `details` properties of an error. Additional error properties may be allowed
through on 4xx and 5xx status code errors using the `safeFields` option to pass
in an array of safe field names:

```ts
app.bind(RestBindings.ERROR_WRITER_OPTIONS).to({safeFields: ['errorCode']});
```

Using the above configuration, an error containing an `errorCode` property will
produce the following response:

```
{
  "error": {
    "statusCode": 500,
    "message": "Internal Server Error",
    "errorCode": "INTERNAL_SERVER_ERROR"
  }
}
```

## Migration from LoopBack 3 error handler

Unlike LoopBack 3’s phase-based middleware routing system, LB4 uses a sequence

> handler that sits infront of a routing table and handles errors. To learn the

Reviewer 4: ```suggestion
handler that sits in front of a routing table and handles errors. To learn the
```

---
title: 'Using strong-error-handler'
lang: en
keywords: LoopBack 4, error handling
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Using-strong-error-handler.html
summary:
  Strong-error-handler is an error handler for use in both development and
  production environments.
---

In LoopBack 4, the request handling process starts with the app's
[sequence](https://loopback.io/doc/en/lb4/Sequence.html), a simple class with
five injected helper methods in the constructor. It is the gatekeeper of all
requests to the app. Errors are handled by one of the Sequence actions
[`reject`](Sequence.md#handling-errors), which calls `strong-error-handler`
package.

# strong-error-handler

This package is an error handler for use in both development (debug) and
production environments.

## Supported versions

| Current | Long Term Support | End-of-Life |
| :-----: | :---------------: | :---------: |

> |   4.x   |        3.x        |     2.x     |

Reviewer 3: We usually format the LTS table differently: each major version is in a new row and we also show the EOL date for each version, see e.g. here:

https://github.com/strongloop/loopback#module-long-term-support-policy

I think there is no need to show 2.x version. The current 3.x version was published at [2018-06-11](https://github.com/strongloop/strong-error-handler/blob/master/CHANGES.md#2018-06-11-version-300), before LB4 GA. I would be very surprised if there was any LB4 user using 2.x version of strong-error handler.

Last but not least, AFAICT, the latest strong-error-handler version is 3.x. I am not sure where is 4.x coming from? See https://github.com/strongloop/strong-error-handler#supported-versions and https://github.com/strongloop/strong-error-handler/blob/master/package.json#L5

Learn more about our LTS plan in
[docs](http://loopback.io/doc/en/contrib/Long-term-support.html).

## Installation

```bash
$ npm install --save strong-error-handler
```

## Use

In production mode, the default implementation of `reject`,
`strong-error-handler` omits details from error responses to prevent leaking

> sensitive information:

Reviewer 3: This section is duplicating existing content in https://loopback.io/doc/en/lb4/Sequence.html#handling-errors. Let's find a way how to avoid that duplication please. For example, the duplicate content on one page can be replaced with a link (a reference) to the other page.
